lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -2356,14 +2360,20 @@ ParseResult FIRStmtParser::parseMapExp(Value &result) {
             parseExp(value, "expected value expression in Map expression"))
           return failure();
 
-        if (key.getType() != keyType)
-          return emitError(loc, "unexpected expression of type ")
-                 << key.getType() << " for key in Map expression, expected "
-                 << keyType;
-        if (value.getType() != valueType)
-          return emitError(loc, "unexpected expression of type ")
-                 << value.getType() << " for value in Map expression, expected "
-                 << valueType;
+        if (key.getType() != keyType) {
+          if (!isa<AnyRefType>(keyType) || !isa<ClassType>(key.getType()))
+            return emitError(loc, "unexpected expression of type ")
+                   << key.getType() << " for key in Map expression, expected "
+                   << keyType;
+          key = builder.create<ObjectAnyRefCastOp>(key);
```

Comment:
I don't think we actually need object references as key types... maybe just remove this part?

Reply:
Oh, hmm.  Moving to reject at FIRRTL parsing (+dialect) level unsupported/illegal combinations of types, probably should be handled separately, which I think is the direction you're suggesting?  That is, supporting the any cast seems reasonable as long as we allow `Map<AnyRef,T>` at all, which we currently do at these levels anyway.  If we want to make these illegal at FIRRTL, rejecting these should be made part of the type (verifier/ODS constraint, like in OM?) and produce nice diagnostic when trying to express such an invalid combination during parsing -- regardless of whether we're constructing one or need any any cast.

Being open by default seems reasonable (esp insofar as FIRRTL properties might want to be more than pre-OM (?)), as does enforcing consistent limitations throughout pipeline esp at point where we can produce nicer diagnostics (earlier too).  Inclined to leave this as-is for this PR, but that's something good to align on especially as these start being written up for the FIRRTL spec proper!  Thanks! :+1: